diff --git a/include/vterm.h b/include/vterm.h
index c6124a1..d63119e 100644
--- a/include/vterm.h
+++ b/include/vterm.h
@@ -233,6 +233,7 @@ typedef union {
 typedef enum {
   /* VTERM_ATTR_NONE = 0 */
   VTERM_ATTR_BOLD = 1,   // bool:   1, 22
+  VTERM_ATTR_FAINT,      // bool:   2, 22
   VTERM_ATTR_UNDERLINE,  // number: 4, 21, 24
   VTERM_ATTR_ITALIC,     // bool:   3, 23
   VTERM_ATTR_BLINK,      // bool:   5, 25
@@ -491,6 +492,7 @@ void vterm_state_send_selection(VTermState *state, VTermSelectionMask mask, VTer
 
 typedef struct {
     unsigned int bold      : 1;
+    unsigned int faint     : 1;
     unsigned int underline : 2;
     unsigned int italic    : 1;
     unsigned int blink     : 1;
@@ -571,19 +573,20 @@ size_t vterm_screen_get_text(const VTermScreen *screen, char *str, size_t len, c
 
 typedef enum {
   VTERM_ATTR_BOLD_MASK       = 1 << 0,
-  VTERM_ATTR_UNDERLINE_MASK  = 1 << 1,
-  VTERM_ATTR_ITALIC_MASK     = 1 << 2,
-  VTERM_ATTR_BLINK_MASK      = 1 << 3,
-  VTERM_ATTR_REVERSE_MASK    = 1 << 4,
-  VTERM_ATTR_STRIKE_MASK     = 1 << 5,
-  VTERM_ATTR_FONT_MASK       = 1 << 6,
-  VTERM_ATTR_FOREGROUND_MASK = 1 << 7,
-  VTERM_ATTR_BACKGROUND_MASK = 1 << 8,
-  VTERM_ATTR_CONCEAL_MASK    = 1 << 9,
-  VTERM_ATTR_SMALL_MASK      = 1 << 10,
-  VTERM_ATTR_BASELINE_MASK   = 1 << 11,
-
-  VTERM_ALL_ATTRS_MASK = (1 << 12) - 1
+  VTERM_ATTR_FAINT_MASK      = 1 << 1,
+  VTERM_ATTR_UNDERLINE_MASK  = 1 << 2,
+  VTERM_ATTR_ITALIC_MASK     = 1 << 3,
+  VTERM_ATTR_BLINK_MASK      = 1 << 4,
+  VTERM_ATTR_REVERSE_MASK    = 1 << 5,
+  VTERM_ATTR_STRIKE_MASK     = 1 << 6,
+  VTERM_ATTR_FONT_MASK       = 1 << 7,
+  VTERM_ATTR_FOREGROUND_MASK = 1 << 8,
+  VTERM_ATTR_BACKGROUND_MASK = 1 << 9,
+  VTERM_ATTR_CONCEAL_MASK    = 1 << 10,
+  VTERM_ATTR_SMALL_MASK      = 1 << 11,
+  VTERM_ATTR_BASELINE_MASK   = 1 << 12,
+
+  VTERM_ALL_ATTRS_MASK = (1 << 13) - 1
 } VTermAttrMask;
 
 int vterm_screen_get_attrs_extent(const VTermScreen *screen, VTermRect *extent, VTermPos pos, VTermAttrMask attrs);
diff --git a/src/pen.c b/src/pen.c
index 704e9d8..5d3105d 100644
--- a/src/pen.c
+++ b/src/pen.c
@@ -170,6 +170,7 @@ INTERNAL void vterm_state_newpen(VTermState *state)
 INTERNAL void vterm_state_resetpen(VTermState *state)
 {
   state->pen.bold = 0;      setpenattr_bool(state, VTERM_ATTR_BOLD, 0);
+  state->pen.faint = 0;     setpenattr_bool(state, VTERM_ATTR_FAINT, 0);
   state->pen.underline = 0; setpenattr_int (state, VTERM_ATTR_UNDERLINE, 0);
   state->pen.italic = 0;    setpenattr_bool(state, VTERM_ATTR_ITALIC, 0);
   state->pen.blink = 0;     setpenattr_bool(state, VTERM_ATTR_BLINK, 0);
@@ -193,6 +194,7 @@ INTERNAL void vterm_state_savepen(VTermState *state, int save)
     state->pen = state->saved.pen;
 
     setpenattr_bool(state, VTERM_ATTR_BOLD,      state->pen.bold);
+    setpenattr_bool(state, VTERM_ATTR_FAINT,     state->pen.faint);
     setpenattr_int (state, VTERM_ATTR_UNDERLINE, state->pen.underline);
     setpenattr_bool(state, VTERM_ATTR_ITALIC,    state->pen.italic);
     setpenattr_bool(state, VTERM_ATTR_BLINK,     state->pen.blink);
@@ -293,12 +295,21 @@ INTERNAL void vterm_state_setpen(VTermState *state, const long args[], int argco
     case 1: { // Bold on
       const VTermColor *fg = &state->pen.fg;
       state->pen.bold = 1;
+      state->pen.faint = 0;
       setpenattr_bool(state, VTERM_ATTR_BOLD, 1);
+      setpenattr_bool(state, VTERM_ATTR_FAINT, 0);
       if(!VTERM_COLOR_IS_DEFAULT_FG(fg) && VTERM_COLOR_IS_INDEXED(fg) && fg->indexed.idx < 8 && state->bold_is_highbright)
         set_pen_col_ansi(state, VTERM_ATTR_FOREGROUND, fg->indexed.idx + (state->pen.bold ? 8 : 0));
       break;
     }
 
+    case 2: // Faint on
+      state->pen.bold = 0;
+      state->pen.faint = 1;
+      setpenattr_bool(state, VTERM_ATTR_BOLD, 0);
+      setpenattr_bool(state, VTERM_ATTR_FAINT, 1);
+      break;
+
     case 3: // Italic on
       state->pen.italic = 1;
       setpenattr_bool(state, VTERM_ATTR_ITALIC, 1);
@@ -357,9 +368,11 @@ INTERNAL void vterm_state_setpen(VTermState *state, const long args[], int argco
       setpenattr_int(state, VTERM_ATTR_UNDERLINE, state->pen.underline);
       break;
 
-    case 22: // Bold off
+    case 22: // Bold and faint off
       state->pen.bold = 0;
+      state->pen.faint = 0;
       setpenattr_bool(state, VTERM_ATTR_BOLD, 0);
+      setpenattr_bool(state, VTERM_ATTR_FAINT, 0);
       break;
 
     case 23: // Italic and Gothic (currently unsupported) off
@@ -506,6 +519,9 @@ INTERNAL int vterm_state_getpen(VTermState *state, long args[], int argcount)
   if(state->pen.bold)
     args[argi++] = 1;
 
+  if(state->pen.faint)
+    args[argi++] = 2;
+
   if(state->pen.italic)
     args[argi++] = 3;
 
@@ -553,6 +569,10 @@ int vterm_state_get_penattr(const VTermState *state, VTermAttr attr, VTermValue
     val->boolean = state->pen.bold;
     return 1;
 
+  case VTERM_ATTR_FAINT:
+    val->boolean = state->pen.faint;
+    return 1;
+
   case VTERM_ATTR_UNDERLINE:
     val->number = state->pen.underline;
     return 1;
diff --git a/src/screen.c b/src/screen.c
index 77c0af8..f0741ba 100644
--- a/src/screen.c
+++ b/src/screen.c
@@ -18,6 +18,7 @@ typedef struct
   VTermColor   fg, bg;
 
   unsigned int bold      : 1;
+  unsigned int faint     : 1;
   unsigned int underline : 2;
   unsigned int italic    : 1;
   unsigned int blink     : 1;
@@ -407,6 +408,9 @@ static int setpenattr(VTermAttr attr, VTermValue *val, void *user)
   case VTERM_ATTR_BOLD:
     screen->pen.bold = val->boolean;
     return 1;
+  case VTERM_ATTR_FAINT:
+    screen->pen.faint = val->boolean;
+    return 1;
   case VTERM_ATTR_UNDERLINE:
     screen->pen.underline = val->number;
     return 1;
@@ -680,6 +684,7 @@ static void resize_buffer(VTermScreen *screen, int bufidx, int new_rows, int new
         }
 
         dst->pen.bold      = src->attrs.bold;
+        dst->pen.faint     = src->attrs.faint;
         dst->pen.underline = src->attrs.underline;
         dst->pen.italic    = src->attrs.italic;
         dst->pen.blink     = src->attrs.blink;
@@ -969,6 +974,7 @@ int vterm_screen_get_cell(const VTermScreen *screen, VTermPos pos, VTermScreenCe
   }
 
   cell->attrs.bold      = intcell->pen.bold;
+  cell->attrs.faint     = intcell->pen.faint;
   cell->attrs.underline = intcell->pen.underline;
   cell->attrs.italic    = intcell->pen.italic;
   cell->attrs.blink     = intcell->pen.blink;
@@ -1086,6 +1092,8 @@ static int attrs_differ(VTermAttrMask attrs, ScreenCell *a, ScreenCell *b)
 {
   if((attrs & VTERM_ATTR_BOLD_MASK)       && (a->pen.bold != b->pen.bold))
     return 1;
+  if((attrs & VTERM_ATTR_FAINT_MASK)      && (a->pen.faint != b->pen.faint))
+    return 1;
   if((attrs & VTERM_ATTR_UNDERLINE_MASK)  && (a->pen.underline != b->pen.underline))
     return 1;
   if((attrs & VTERM_ATTR_ITALIC_MASK)     && (a->pen.italic != b->pen.italic))
diff --git a/src/vterm.c b/src/vterm.c
index b2f61d2..5481651 100644
--- a/src/vterm.c
+++ b/src/vterm.c
@@ -265,6 +265,7 @@ VTermValueType vterm_get_attr_type(VTermAttr attr)
 {
   switch(attr) {
     case VTERM_ATTR_BOLD:       return VTERM_VALUETYPE_BOOL;
+    case VTERM_ATTR_FAINT:      return VTERM_VALUETYPE_BOOL;
     case VTERM_ATTR_UNDERLINE:  return VTERM_VALUETYPE_INT;
     case VTERM_ATTR_ITALIC:     return VTERM_VALUETYPE_BOOL;
     case VTERM_ATTR_BLINK:      return VTERM_VALUETYPE_BOOL;
diff --git a/src/vterm_internal.h b/src/vterm_internal.h
index 6aa9007..d2a56f2 100644
--- a/src/vterm_internal.h
+++ b/src/vterm_internal.h
@@ -41,6 +41,7 @@ struct VTermPen
   VTermColor fg;
   VTermColor bg;
   unsigned int bold:1;
+  unsigned int faint:1;
   unsigned int underline:2;
   unsigned int italic:1;
   unsigned int blink:1;
